broker-core/src/main/java/io/zeebe/broker/workflow/processor/deployment/DeploymentCreatedProcessor.java

```diff
@@ -55,32 +57,50 @@ public void processRecord(
     }
 
     for (final Workflow workflowRecord : deploymentEvent.workflows()) {
-      if (workflowRecord.getVersion() != 1) {
-        closeExistingMessageStartEventSubscriptions(workflowState, workflowRecord, streamWriter);
+      if (!isLatestWorkflow(workflowRecord)) {
+        continue;
       }
+
+      closeExistingMessageStartEventSubscriptions(workflowState, workflowRecord, streamWriter);
       openMessageStartEventSubscriptions(workflowState, workflowRecord, streamWriter);
     }
   }
 
+  private boolean isLatestWorkflow(Workflow workflow) {
+    return workflowState
+            .getLatestWorkflowVersionByProcessId(workflow.getBpmnProcessId())
+            .getVersion()
+        == workflow.getVersion();
+  }
+
   private void closeExistingMessageStartEventSubscriptions(
       WorkflowState workflowState, Workflow workflowRecord, TypedStreamWriter streamWriter) {
-    final int previousVersion = workflowRecord.getVersion() - 1;
-    final DeployedWorkflow previousWorkflow =
-        workflowState.getWorkflowByProcessIdAndVersion(
-            workflowRecord.getBpmnProcessId(), previousVersion);
-    if (previousWorkflow == null
-        || previousWorkflow.getWorkflow().getStartEvents().stream().noneMatch(e -> e.isMessage())) {
+    final DeployedWorkflow lastMsgWorkflow = findLastMessageStartWorkflow(workflowRecord);
+    if (lastMsgWorkflow == null) {
       return;
     }
 
-    final long previousWorkflowKey = previousWorkflow.getKey();
-    final MessageStartEventSubscriptionRecord subscriptionRecord =
-        new MessageStartEventSubscriptionRecord();
-    subscriptionRecord.setWorkflowKey(previousWorkflowKey);
-
+    subscriptionRecord.reset();
+    subscriptionRecord.setWorkflowKey(lastMsgWorkflow.getKey());
     streamWriter.appendNewCommand(MessageStartEventSubscriptionIntent.CLOSE, subscriptionRecord);
   }
 
+  private DeployedWorkflow findLastMessageStartWorkflow(final Workflow workflowRecord) {
+    DeployedWorkflow lastMsgWorkflow;
+
+    for (int version = workflowRecord.getVersion() - 1; version > 0; --version) {
+      lastMsgWorkflow =
+          workflowState.getWorkflowByProcessIdAndVersion(
+              workflowRecord.getBpmnProcessId(), version);
+      if (lastMsgWorkflow != null
+          && lastMsgWorkflow.getWorkflow().getStartEvents().stream().anyMatch(e -> e.isMessage())) {
+        return lastMsgWorkflow;
+      }
+    }
+
+    return null;
+  }
+
   private void openMessageStartEventSubscriptions(
       WorkflowState workflowState, Workflow workflowRecord, TypedStreamWriter streamWriter) {
```

Comment:
We are currently only close one subscription but open a subscription for each message start event. I think we should also close for each message start event. You could re-use this logic to close all open message start events.

Reply:
We only issue one CLOSE command but the close processor removes all subscriptions of that workflow. I think it's because when you open a subscription, you open it for a specific start event, but when you close you want all of them to close.